modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/defragmentation/IndexingDefragmentation.java

```diff
@@ -0,0 +1,467 @@
+import static org.apache.ignite.internal.processors.query.h2.defragmentation.IndexingDefragmentation.IndexStages.READ_ROW;
+
+/**
+ *
+ */
+public class IndexingDefragmentation {
+    /** Indexing. */
+    private final IgniteH2Indexing indexing;
+
+    /** Constructor. */
+    public IndexingDefragmentation(IgniteH2Indexing indexing) {
+        this.indexing = indexing;
+    }
+
+    /** */
+    public enum IndexStages {
+        START,
+        CP_LOCK,
+        INIT_TREE,
+        ITERATE,
+        READ_ROW,
+        READ_MAP,
+        INSERT_ROW
+    }
+
+    /**
+     * Defragment index partition.
+     *
+     * @param grpCtx Old group context.
+     * @param newCtx New group context.
+     * @param partPageMem Partition page memory.
+     * @param mappingByPartition Mapping page memory.
+     * @param cpLock Defragmentation checkpoint read lock.
+     * @param log Log.
+     *
+     * @throws IgniteCheckedException If failed.
+     */
+    public void defragment(
+        CacheGroupContext grpCtx,
+        CacheGroupContext newCtx,
+        PageMemoryEx partPageMem,
+        IntMap<LinkMap> mappingByPartition,
+        CheckpointTimeoutLock cpLock,
+        IgniteLogger log
+    ) throws IgniteCheckedException {
+        int pageSize = grpCtx.cacheObjectContext().kernalContext().grid().configuration().getDataStorageConfiguration().getPageSize();
+
+        TreeIterator treeIterator = new TreeIterator(pageSize);
+
+        PageMemoryEx oldCachePageMem = (PageMemoryEx)grpCtx.dataRegion().pageMemory();
+
+        PageMemory newCachePageMemory = partPageMem;
+
+        Collection<GridH2Table> tables = indexing.schemaManager().dataTables();
+
+        long cpLockThreshold = 150L;
+
+        TimeTracker<IndexStages> tracker = new TimeTracker<>(IndexStages.class);
+
+        cpLock.checkpointReadLock();
+        tracker.complete(CP_LOCK);
+
+        try {
+            AtomicLong lastCpLockTs = new AtomicLong(System.currentTimeMillis());
+
+            for (GridH2Table table : tables) {
+                GridCacheContext<?, ?> cctx = table.cacheContext();
+
+                if (cctx.groupId() != grpCtx.groupId())
+                    continue; // Not our index.
+
+                GridH2RowDescriptor rowDesc = table.rowDescriptor();
+
+                List<Index> indexes = table.getIndexes();
+                H2TreeIndex oldH2Idx = (H2TreeIndex)indexes.get(2);
+
+                int segments = oldH2Idx.segmentsCount();
+
+                H2Tree firstTree = oldH2Idx.treeForRead(0);
+
+                PageIoResolver pageIoRslvr = pageAddr -> {
+                    PageIO io = PageIoResolver.DEFAULT_PAGE_IO_RESOLVER.resolve(pageAddr);
+
+                    if (io instanceof BPlusMetaIO)
+                        return io;
+
+                    //noinspection unchecked,rawtypes,rawtypes
+                    return wrap((BPlusIO)io);
+                };
+
+                //TODO Create new proper GridCacheContext for it?
```

Comment:
> Please, remove it.

Please show me how do I use TimeBag for the same scenario. I'm sure that I can't do that.

> Let's move everything here from PageStoreMap here? The DefragmentationPageReadWriteManager will also implement the PageStoreCollection.

No, DefragmentationPageReadWriteManager is instantiated later then we need page stores map. Why would I put this code inside of the manager? Reason #2 - we have two maps, not one.

> It will be simpler and better if you'll reuse try-finally approach for all these cases.

Ok, I'm removing all this code.

Reply:
Done